crates/common/download/src/download.rs

```diff
@@ -225,70 +330,145 @@ impl Downloader {
         }
     }
 
+    /// Deletes the file if it was downloaded.
     pub async fn cleanup(&self) -> Result<(), DownloadError> {
         let _res = tokio::fs::remove_file(&self.target_filename).await;
         Ok(())
     }
 }
 
+/// Requests either the entire HTTP resource, or its part, from an offset to the
+/// end.
+///
+/// If `range_start` is `0`, then a regular GET request is sent. Otherwise, a
+/// request for a range of the resource, starting from `range_start`, until EOF,
+/// is sent.
+///
+/// We use a half-open range with only a lower bound, because we expect to use
+/// it to download static resources which do not change, and only as a recovery
+/// mechanism in case of network failures.
+async fn request_range_from(
+    url: &DownloadInfo,
+    range_start: u64,
+) -> Result<reqwest::Response, reqwest::Error> {
+    // Default retry is an exponential retry with a limit of 15 minutes total.
+    // Let's set some more reasonable retry policy so we don't block the downloads for too long.
+    let backoff = ExponentialBackoff {
+        initial_interval: BACKOFF_INITIAL_INTERVAL,
+        max_elapsed_time: Some(BACKOFF_MAX_ELAPSED),
+        ..Default::default()
+    };
+
+    let operation = || async {
+        let mut client = reqwest::Client::new().get(url.url());
+
+        if let Some(Auth::Bearer(token)) = &url.auth {
+            client = client.bearer_auth(token)
+        }
+
+        if range_start != 0 {
+            client = client.header("Range", format!("bytes={range_start}-"));
+        }
+
+        client
+            .send()
+            .await
+            .map_err(|err| {
+                if err.is_connect() || err.is_builder() {
+                    backoff::Error::Permanent(err)
+                } else {
+                    log::warn!("Failed to Download. {:?}\nRetrying.", &err);
+                    backoff::Error::transient(err)
+                }
+            })?
+            .error_for_status()
+            .map_err(|err| match err.status() {
+                Some(status_error) if status_error.is_client_error() => {
+                    backoff::Error::Permanent(err)
+                }
+                _ => backoff::Error::transient(err),
+            })
+    };
+
+    retry(backoff, operation).await
+}
+
+/// Saves a response body chunks starting from an offset.
+async fn save_chunks_to_file_at(
+    response: &mut reqwest::Response,
+    writer: &mut File,
+    offset: u64,
+) -> Result<(), SaveChunksError> {
```

Comment:
This is a really neat example of how a function specific error type really helps to improve error handling!

Here to keep trying on network error and give up on disk error.